File: HackRfTcp/fm_demodulator.py
import socket
import numpy as np
import sounddevice as sd
from scipy import signal
import time
import threading
from pynput import keyboard

# HackRF Configuration
HACKRF_IP = '192.168.1.2'
CONTROL_PORT = 5001
DATA_PORT = 5000

FREQUENCY = 100000000      # 100 MHz
SAMPLE_RATE = 2000000      # 2 MHz
AUDIO_RATE = 48000
VGA_GAIN = 30
LNA_GAIN = 32
RX_AMP_GAIN = 14

FREQ_STEP = 100000  # 100 kHz

# Frequency control
current_freq = FREQUENCY
freq_change_active = {'left': False, 'right': False}
freq_thread_running = True

def control_hackrf(host, port, command):
    """Send command to HackRF control port"""
    s = socket.socket()
    s.connect((host, port))
    
    # Read welcome message on first connection
    welcome = s.recv(4096).decode()
    
    # Send command
    s.send(command.encode() + b'\n')
    response = s.recv(1024).decode()
    s.close()
    
    # Return only relevant part
    return "OK" if "Ready" in response else response.split('\n')[0]

def set_frequency(freq):
    """Set new frequency"""
    control_hackrf(HACKRF_IP, CONTROL_PORT, f'SET_FREQ:{freq}')
    print(f"\r>>> Frequency: {freq/1e6:.3f} MHz     ", end='', flush=True)
    return freq

def frequency_control_thread():
    """Background thread for continuous frequency changes"""
    global current_freq
    while freq_thread_running:
        if freq_change_active['left']:
            current_freq -= FREQ_STEP
            set_frequency(current_freq)
            time.sleep(0.1)
        elif freq_change_active['right']:
            current_freq += FREQ_STEP
            set_frequency(current_freq)
            time.sleep(0.1)
        else:
            time.sleep(0.05)

def on_press(key):
    """Handle key press"""
    try:
        if key == keyboard.Key.left:
            freq_change_active['left'] = True
        elif key == keyboard.Key.right:
            freq_change_active['right'] = True
    except:
        pass

def on_release(key):
    """Handle key release"""
    try:
        if key == keyboard.Key.left:
            freq_change_active['left'] = False
        elif key == keyboard.Key.right:
            freq_change_active['right'] = False
    except:
        pass

def fm_demod(iq_samples):
    """Demodulate FM signal"""
    i = iq_samples[0::2].astype(np.float32) / 127.0
    q = iq_samples[1::2].astype(np.float32) / 127.0
    iq = i + 1j * q
    phase = np.angle(iq)
    demod = np.diff(np.unwrap(phase))
    return demod

def deemphasis_filter(audio_rate):
    """75µs de-emphasis filter for broadcast FM"""
    d = audio_rate * 75e-6
    x = np.exp(-1 / d)
    b = [1 - x]
    a = [1, -x]
    return b, a

def receive_iq_data():
    """Main function to receive and process IQ data"""
    global current_freq
    
    # Configure HackRF
    print("Configuring HackRF...")
    control_hackrf(HACKRF_IP, CONTROL_PORT, f'SET_FREQ:{FREQUENCY}')
    control_hackrf(HACKRF_IP, CONTROL_PORT, f'SET_SAMPLE_RATE:{SAMPLE_RATE}')
    control_hackrf(HACKRF_IP, CONTROL_PORT, f'SET_VGA_GAIN:{VGA_GAIN}')
    control_hackrf(HACKRF_IP, CONTROL_PORT, f'SET_LNA_GAIN:{LNA_GAIN}')
    control_hackrf(HACKRF_IP, CONTROL_PORT, f'SET_RX_AMP_GAIN:{RX_AMP_GAIN}')
    
    # Calculate decimation
    decimation = int(SAMPLE_RATE / AUDIO_RATE)
    print(f"Decimation factor: {decimation}")
    
    # Setup de-emphasis filter
    b, a = deemphasis_filter(AUDIO_RATE)
    zi = signal.lfilter_zi(b, a)
    
    # Connect to data stream
    print("Connecting to data stream...")
    data_sock = socket.socket()
    data_sock.connect((HACKRF_IP, DATA_PORT))
    
    # Setup audio output
    stream = sd.OutputStream(samplerate=AUDIO_RATE, channels=1, dtype='float32')
    stream.start()
    
    # Start keyboard listener
    listener = keyboard.Listener(on_press=on_press, on_release=on_release)
    listener.start()
    
    # Start frequency control thread
    freq_thread = threading.Thread(target=frequency_control_thread, daemon=True)
    freq_thread.start()
    
    print("\n=== FM Demodulator Running ===")
    print(f"Frequency: {FREQUENCY/1e6:.3f} MHz")
    print(f"Sample Rate: {SAMPLE_RATE/1e6:.2f} MHz")
    print("Controls: LEFT/RIGHT arrows to change frequency")
    print("Press Ctrl+C to stop.\n")
    
    total_samples = 0
    try:
        while True:
            # Receive IQ data
            data = data_sock.recv(262144)
            iq = np.frombuffer(data, dtype=np.int8)
            
            if len(iq) == 0:
                continue
            
            total_samples += len(iq)
            
            # Demodulate FM
            audio = fm_demod(iq)
            
            # Decimate to audio rate
            if decimation > 1:
                audio = signal.decimate(audio, decimation, zero_phase=True, ftype='fir')
            
            # Apply de-emphasis
            audio, zi = signal.lfilter(b, a, audio, zi=zi)
            
            # Normalize
            max_val = np.max(np.abs(audio))
            if max_val > 0:
                audio = audio / max_val * 0.5
            
            # Play audio
            stream.write(audio.astype(np.float32))
            
    except KeyboardInterrupt:
        print(f"\n\nTotal samples received: {total_samples:,}")
        print("Stopping...")
    finally:
        global freq_thread_running
        freq_thread_running = False
        listener.stop()
        stream.stop()
        stream.close()
        data_sock.close()
        print("Closed.")

# Run the receiver
if __name__ == "__main__":
    receive_iq_data()

File: HackRFReceiverIOS/HackRFReceiverIOS/FMDemodulator.swift
import Foundation
import Accelerate

class FMDemodulator {
    private let sampleRate: Int
    private let audioRate: Int
    private let decimation: Int
    
    // De-emphasis filter coefficients (75µs for broadcast FM)
    private var filterB: [Double] = []
    private var filterA: [Double] = []
    
    init(sampleRate: Int, audioRate: Int = 48000) {
        self.sampleRate = sampleRate
        self.audioRate = audioRate
        self.decimation = max(1, sampleRate / audioRate)
        
        setupDeemphasisFilter()
    }
    
    private func setupDeemphasisFilter() {
        // 75µs de-emphasis filter for broadcast FM
        let d = Double(audioRate) * 75e-6
        let x = exp(-1.0 / d)
        
        filterB = [1.0 - x]
        filterA = [1.0, -x]
    }
    
    func demodulate(iqSamples: [Int8]) -> [Double]? {
        guard iqSamples.count >= 4 else { return nil }
        
        // Convert to I/Q
        var i = [Double]()
        var q = [Double]()
        
        for idx in stride(from: 0, to: iqSamples.count - 1, by: 2) {
            i.append(Double(iqSamples[idx]) / 127.0)
            q.append(Double(iqSamples[idx + 1]) / 127.0)
        }
        
        guard i.count > 1 else { return nil }
        
        // Calculate phase
        var phases = [Double](repeating: 0.0, count: i.count)
        for idx in 0..<i.count {
            phases[idx] = atan2(q[idx], i[idx])
        }
        
        // Unwrap phase
        phases = unwrapPhase(phases)
        
        // Phase differentiation (FM demod)
        var demod = [Double](repeating: 0.0, count: phases.count - 1)
        for idx in 0..<demod.count {
            demod[idx] = phases[idx + 1] - phases[idx]
        }
        
        guard !demod.isEmpty else { return nil }
        
        // Decimate to audio rate
        if decimation > 1 {
            demod = decimate(demod, factor: decimation)
        }
        
        // Apply de-emphasis
        demod = applyDeemphasis(demod)
        
        // Normalize
        if let maxVal = demod.max(by: { abs($0) < abs($1) }), abs(maxVal) > 0.01 {
            let scale = 0.5 / abs(maxVal)
            demod = demod.map { $0 * scale }
        }
        
        return demod
    }
    
    private func unwrapPhase(_ phases: [Double]) -> [Double] {
        var unwrapped = [Double](repeating: 0.0, count: phases.count)
        var correction = 0.0
        
        unwrapped[0] = phases[0]
        
        for i in 1..<phases.count {
            var diff = phases[i] - phases[i-1]
            
            while diff > .pi {
                diff -= 2.0 * .pi
                correction -= 2.0 * .pi
            }
            while diff < -.pi {
                diff += 2.0 * .pi
                correction += 2.0 * .pi
            }
            
            unwrapped[i] = phases[i] + correction
        }
        
        return unwrapped
    }
    
    private func decimate(_ signal: [Double], factor: Int) -> [Double] {
        guard factor > 1 else { return signal }
        
        let filterLength = min(31, signal.count / 4)
        let cutoff = 0.4 / Double(factor)
        let firCoeffs = designLowpassFIR(length: filterLength, cutoff: cutoff)
        
        var filtered = [Double](repeating: 0.0, count: signal.count)
        
        for i in 0..<signal.count {
            var sum = 0.0
            for j in 0..<filterLength {
                let idx = i - j + filterLength / 2
                if idx >= 0 && idx < signal.count {
                    sum += signal[idx] * firCoeffs[j]
                }
            }
            filtered[i] = sum
        }
        
        var decimated = [Double]()
        for i in stride(from: 0, to: filtered.count, by: factor) {
            decimated.append(filtered[i])
        }
        
        return decimated
    }
    
    private func designLowpassFIR(length: Int, cutoff: Double) -> [Double] {
        var coeffs = [Double](repeating: 0.0, count: length)
        let center = Double(length - 1) / 2.0
        
        for i in 0..<length {
            let x = Double(i) - center
            if abs(x) < 0.001 {
                coeffs[i] = 2.0 * cutoff
            } else {
                coeffs[i] = sin(2.0 * .pi * cutoff * x) / (.pi * x)
            }
            
            // Hamming window
            coeffs[i] *= 0.54 - 0.46 * cos(2.0 * .pi * Double(i) / Double(length - 1))
        }
        
        let sum = coeffs.reduce(0, +)
        if abs(sum) > 0.001 {
            coeffs = coeffs.map { $0 / sum }
        }
        
        return coeffs
    }
    
    private func applyDeemphasis(_ signal: [Double]) -> [Double] {
        guard !signal.isEmpty else { return [] }
        
        var output = [Double](repeating: 0.0, count: signal.count)
        
        for i in 0..<signal.count {
            var y = filterB[0] * signal[i]
            
            if i > 0 {
                y -= filterA[1] * output[i-1]
            }
            
            output[i] = y
        }
        
        return output
    }
}
